### ISSUE TITLE
About biLSTM model  different sequence lenght use padding & mask.

### ISSUE THREAD

**Issue by BeccaHuang** [tensorflow/tensorflow#48702]
About 「Variable Length Sequence Support for CUDNN LSTM #23269」

i have three question about biLSTM model have different sequence lenght that i will use padding & mask.

>In 2.0 alpha, the normal keras.layer.LSTM will use cudnn kernel if possible, so you don't need to use keras.layers.CuDNNLSTM layer any more.

(1)So in tf2 I can use lstm  don't need to use keras.layers.CuDNNLSTM layer,but why is the GPU running speed different tf2 LSTM slow then tf1 CuDNNLSTM?

> The cudnn kernel only support seqence_length parameter, so that it can skip the padding data on the tail. If your data has padding at the beginning of the sequence (left padded), then cudnn kernel will not be able to work correctly in this situation. The comment here is trying to point out that specific use case.

 (2)So if I use  tf2  layer→ tf.keras.layers.Bidirectional(tf.compat.v1.keras.layers.CuDNNLSTM ....),input data already padding on the tail,then cudnn kernel will be able to work correctly in this situation? But how can i prove it about masked,about model runing loss not consider padded location  ? 

(3)If padding data then model output result I need to drop padding location  right?

**Comment by tilakrayal** [comment]
@BeccaHuang ,

In order to expedite the trouble-shooting process, could you please provide the TensorFlow version,the  complete code and the dataset you are using. Thanks!

**Comment by BeccaHuang** [comment]
Hi @tilakrayal sorry i can't provide dataset but model i build,

def build_LSTM_model(input_shape): 
    inp =tf.keras.layers.Input(shape=(input_shape[1], input_shape[2]))
    encoder = tf.keras.layers.Bidirectional(tf.compat.v1.keras.layers.CuDNNLSTM(128, activity_regularizer=regularizers.l1(10e-9),return_sequences=True), name='en_Bi_LSTM_1')(inp)
    encoder_output = tf.keras.layers.Bidirectional(tf.compat.v1.keras.layers.CuDNNLSTM(5,activity_regularizer=regularizers.l1(10e-9),return_sequences=False),name='en_Bi_LSTM_3')(encoder)
    encoded_layers = layers.Dense(MAX_LENGTH*10, activation='relu')(encoder_output)
    context_vector = layers.Reshape((MAX_LENGTH,10), input_shape=(MAX_LENGTH*10,))(encoded_layers)
    decoder = tf.keras.layers.Bidirectional(tf.compat.v1.keras.layers.CuDNNLSTM(5,return_sequences=True), name='de_Bi_LSTM_1')(context_vector)
    decoder= tf.keras.layers.Bidirectional(tf.compat.v1.keras.layers.CuDNNLSTM(128,return_sequences=True),name='de_Bi_LSTM_1')(decoder)
     out_p = tf.keras.layers.TimeDistributed(tf.keras.layers.Dense(input_shape[2], activation='linear', name='Dense_1'))(decoder)

Sorry ,I don't know how more correct display my question.

**Comment by tilakrayal** [comment]
@BeccaHuang ,

Please share all the dependencies to replicate the issue or share a colab gist with the reported error.

Thanks!

**Comment by BeccaHuang** [comment]
Thank you Tilakrayal. 
I don't have error about code,but i just want to know if add mask layer how to run in model....
about Variable Length Sequence Support for CUDNN LSTM #23269
in #23269 say if padding in tail  will be able to work correctly in this situation?

**Comment by ymodak** [comment]
>  In 2.0 alpha, the normal keras.layer.LSTM will use cudnn kernel if possible, so you don't need to use keras.layers.CuDNNLSTM layer any more.
 (1)So in tf2 I can use lstm don't need to use keras.layers.CuDNNLSTM layer,but why is the GPU running speed different tf2 LSTM slow then tf1 CuDNNLSTM?

For above question, you need to make sure the configuration is correct for using `tf.keras.layer.LSTM` layer. Since you are experiencing slower performance we should first verify if the gpu setup is done correctly.
Make sure you have compatible cuda, cudnn and nvidia versions installed and the system recognizes gpu.
You may refer this [table](https://www.tensorflow.org/install/source#gpu) to know more.
If you can test with later TF versions such as 2.4 or 2.5 that will be great since there are high chances that performance must have improved since 2.0 alpha.

**Comment by BeccaHuang** [comment]
Hi @ymodak,
thanks you vary much about you reply
I will print("GPU Available: ", tf.test.is_gpu_available()) and  print(tf.__version__) to check

GPU Available:  True
tf version: 2.3.1
cuda:11.1

i need to change version? or i need to check my model layer,
but i run the same layer in tf1 cudnnLSTM  and tf2 LSTM run speed different tf1 is better.

**Comment by BeccaHuang** [comment]
Hi @ymodak,
thanks you vary much about you reply,
although my system version does not match but test GPU available: TRUE,
see the system still use GPU run. Still compare I recommend that I use the corresponding version?
or can run basically effectiveness does not differ too much?
